workflow/metrics/collector.go

```diff
@@ -131,4 +156,77 @@ func (wc *workflowCollector) collectWorkflow(ch chan<- prometheus.Metric, wf wfv
 		addGauge(descWorkflowFinishedAt, float64(wf.Status.FinishedAt.Unix()))
 	}
 
+	// Collect Node metrics
+	nodes := wf.Status.Nodes
+	if wf.Status.IsOffloadNodeStatus() && wc.offloadNodeStatusRepoSupplier != nil {
+		offloadNodeStatusRepo := wc.offloadNodeStatusRepoSupplier.GetOffloadNodeStatusRepo()
+		if offloadNodeStatusRepo.IsEnabled() {
+			offloadedNodes, err := offloadNodeStatusRepo.Get(string(wf.UID), wf.GetOffloadNodeStatusVersion())
+			if err != nil {
+				log.Errorf("not able to get offloaded nodes")
+			}
+			nodes = offloadedNodes
+		}
+	}
+
+	for _, node := range nodes {
+		wc.collectWorkflowNode(ch, node, wf.Name, wf.Namespace)
+	}
+}
+
+func (wc *workflowCollector) collectWorkflowNode(ch chan<- prometheus.Metric, node wfv1.NodeStatus, wfName, wfNamespace string) {
+	addConstMetric := func(desc *prometheus.Desc, t prometheus.ValueType, v float64, lv ...string) {
+		lv = append([]string{wfNamespace, wfName, node.Name}, lv...)
+		ch <- prometheus.MustNewConstMetric(desc, t, v, lv...)
+	}
+	addGauge := func(desc *prometheus.Desc, v float64, lv ...string) {
+		addConstMetric(desc, prometheus.GaugeValue, v, lv...)
+	}
+
+	addGauge(descWorkflowNodeStatusPhase, boolFloat64(node.Phase == wfv1.NodePending || node.Phase == ""), string(wfv1.NodePending))
+	addGauge(descWorkflowNodeStatusPhase, boolFloat64(node.Phase == wfv1.NodeRunning), string(wfv1.NodeRunning))
+	addGauge(descWorkflowNodeStatusPhase, boolFloat64(node.Phase == wfv1.NodeSucceeded), string(wfv1.NodeSucceeded))
+	addGauge(descWorkflowNodeStatusPhase, boolFloat64(node.Phase == wfv1.NodeSkipped), string(wfv1.NodeSkipped))
+	addGauge(descWorkflowNodeStatusPhase, boolFloat64(node.Phase == wfv1.NodeFailed), string(wfv1.NodeFailed))
+	addGauge(descWorkflowNodeStatusPhase, boolFloat64(node.Phase == wfv1.NodeError), string(wfv1.NodeError))
+
+	if !node.StartedAt.IsZero() {
+		addGauge(descWorkflowNodeStartedAt, float64(node.StartedAt.Unix()))
+	}
+
+	if !node.FinishedAt.IsZero() {
+		addGauge(descWorkflowNodeFinishedAt, float64(node.FinishedAt.Unix()))
+	}
+
+	if node.Outputs != nil {
+		for _, param := range node.Outputs.Parameters {
+			if param.EmitMetric != nil {
+				emitMetricOpts := param.EmitMetric
+
+				metricName := metricPrefix + emitMetricOpts.MetricSuffix
+				metricDescription := fmt.Sprintf("Custom metric '%s' from Workflow '%s'", param.Name, wfName)
+				metricLabelNames := descWorkflowStepDefaultLabels
+				var metricLabelValues []string
+
+				for _, metricTag := range emitMetricOpts.MetricTags {
+					metricLabelNames = append(metricLabelNames, metricTag.Name)
+					metricLabelValues = append(metricLabelValues, metricTag.Value)
+				}
+
+				metricDesc := prometheus.NewDesc(
+					metricName,
+					metricDescription,
+					metricLabelNames,
+					nil,
+				)
+
+				parsedValue, err := strconv.ParseFloat(*param.Value, 64)
+				if err == nil {
```

Comment:
please flip this condition so it is conventional - and then make sure we (a) log the error and (b) log it at error level